### PR TITLE
Install and config for GSI indexes

### LINKED ISSUE #3353: GSI - Install and upgrade
Installation will follow the same approach implemented in 2.0 for view install/upgrade:
 - Index names will include an index version suffix
 - On Sync Gateway startup, it will check for existence of required index versions, and create if not present
 - If the CreateIndex returns an error indicating that there are no index services available in the cluster, Sync Gateway startup should fail with an appropriate error message.
 - As part of the existence check, SG will also check if num_replica for the existing indexes matches the value specified in SG’s config.  If not, will drop and recreate the index.  Note: Need to consider how to avoid a scenario where different SG nodes specify different values for num_replica, and get stuck in drop/create battle. 
 - On Sync Gateway startup, will verify index readiness (via query) before starting Sync Gateway’s REST API
 - /db/_post_upgrade endpoint is used to remove obsolete index definitions, and is run after SG cluster upgrade is complete

### PR DESCRIPTION
**Pull request by adamcfraser**
Adds startup installation of GSI indexes when not already present, including readiness checks and upgrade handling.

Views are still created and used at runtime as usual - functional changes to switch to view usage included in subsequent change.

Fixes #3353.

### CODE REVIEW

File: db/indexes.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+			if err != nil {
+				return removedIndexes, err
+			}
+		}
+	}
+	return removedIndexes, nil
+}
+
+// Removes an obsolete index from the database.  In preview mode, checks for existence of the index only.
+func removeObsoleteIndex(bucket *base.CouchbaseBucketGoCB, removedIndexes []string, indexName string, previewOnly bool) (removedIndexesOut []string, err error) {
```
Comment (bbrks) [inline db/indexes.go]: Appending to a slice that gets passed in feels a little awkward to me. I can see its use in `removeObsoleteIndexes` but I think that should be managing its own slice, and have this function return a `removed bool` instead.

```go
func removeObsoleteIndex(bucket *base.CouchbaseBucketGoCB, indexName string, previewOnly bool) (removed bool, err error)
```

**Review by bbrks — COMMENTED** [review]
Left a few comments, but in general it looks good.

File: db/indexes.go (hunk 1 above)
Comment (adamcfraser) [reply db/indexes.go]: Good call.  It ends up with some additional duplicated code in removeObsoleteIndexes (multiple places need to check for the removed flag), but it's a more appropriate functional separation.

**Comment by adamcfraser** [comment]
@bbrks Made some tweaks based on your feedback - have a look.

**Review by bbrks — APPROVED** [review]
Nice, new changes look good!

**Pull request merged**